velox/functions/sparksql/SplitFunctions.cpp

```diff
@@ -16,124 +16,240 @@
 
 #include <utility>
 
+#include "velox/expression/DecodedArgs.h"
 #include "velox/expression/VectorFunction.h"
 #include "velox/expression/VectorWriters.h"
+#include "velox/functions/lib/Re2Functions.h"
+#include "velox/functions/prestosql/Utf8Utils.h"
 
 namespace facebook::velox::functions::sparksql {
 namespace {
 
-/// This class only implements the basic split version in which the pattern is a
-/// single character
-class SplitCharacter final : public exec::VectorFunction {
+// split(string, delimiter[, limit]) -> array(varchar)
+//
+// Splits string on delimiter and returns an array of size at most limit.
+// delimiter is a string representing regular expression.
+// limit is an integer which controls the number of times the regex is applied.
+// By default, limit is -1.
+class Split final : public exec::VectorFunction {
  public:
-  explicit SplitCharacter(const char pattern) : pattern_{pattern} {
-    static constexpr std::string_view kRegexChars = ".$|()[{^?*+\\";
-    VELOX_CHECK(
-        kRegexChars.find(pattern) == std::string::npos,
-        "This version of split supports single-length non-regex patterns");
-  }
+  Split() {}
 
   void apply(
       const SelectivityVector& rows,
       std::vector<VectorPtr>& args,
       const TypePtr& /* outputType */,
       exec::EvalCtx& context,
       VectorPtr& result) const override {
-    exec::LocalDecodedVector input(context, *args[0], rows);
+    exec::DecodedArgs decodedArgs(rows, args, context);
+    DecodedVector* strings = decodedArgs.at(0);
+    DecodedVector* delims = decodedArgs.at(1);
+    DecodedVector* limits = (args.size() == 2) ? nullptr : decodedArgs.at(2);
 
     BaseVector::ensureWritable(rows, ARRAY(VARCHAR()), context.pool(), result);
     exec::VectorWriter<Array<Varchar>> resultWriter;
     resultWriter.init(*result->as<ArrayVector>());
 
-    rows.applyToSelected([&](vector_size_t row) {
-      resultWriter.setOffset(row);
-      auto& arrayWriter = resultWriter.current();
-
-      const StringView& current = input->valueAt<StringView>(row);
-      const char* pos = current.begin();
-      const char* end = pos + current.size();
-      const char* delim;
-      do {
-        delim = std::find(pos, end, pattern_);
-        arrayWriter.add_item().setNoCopy(StringView(pos, delim - pos));
-        pos = delim + 1; // Skip past delim.
-      } while (delim != end);
+    auto getLimit = [&](vector_size_t row) -> int32_t {
+      int32_t limit = std::numeric_limits<int32_t>::max();
+      if (limits) {
+        const auto limitValue = limits->valueAt<vector_size_t>(row);
```

Comment:
use `valueAt<int32_t>` because limit is of int32_t type. We just changed to use vector_size_t for the index.

Reply:
sure thing, my mistake